### DISCUSSION TITLE
sticker segmentation

### DISCUSSION THREAD

**Discussion by mwarning** [ImageMagick/ImageMagick#2537] [Help]
Hi,

I like to segment photographs of multiple stickers into single images.
Here is an example image: https://github.com/securingdev/awesome-stickers/blob/master/raw/attrition_01.pdf

The current approach I tried is one on [stackoverflow](https://stackoverflow.com/a/29758578/2445758). It works well for the example there. Probably because the background is solid white.

For the image with stickers, it does not work at all even when I fiddle with the options.
Cutting out the background works rather well in gimp (fuzzy select), but I like to automate it with ImageMagick.

Maybe someone can give me some hints on how to cut out the background with ImageMagick, so I can finish my segmentation script.

**Comment by fmw42** [comment]
Here is one way to do that. 

```
I homogenize the background color
Then reduce colors to 32
Then apply morphology close
Then convert all colors to white except that background and the background to black
Then use connected components to filter out small regions and get the bounding boxes from the binary image
Then loop over all the bounding boxes and crop the input image at each bounding box

```

Unix syntax:

```
cd
cd desktop/test1
bboxArr=(`convert attrition_01.pdf +repage -alpha off \
-fuzz 15% -fill "rgb(235,193,134)" -opaque "rgb(235,193,134)" \
+dither -colors 32   \
-morphology close square:1 \
-fuzz 0 -fill white +opaque "rgb(235,193,134)" -fill black +opaque white \
-type bilevel \
-define connected-components:exclude-header=true \
-define connected-components:verbose=true \
-define connected-components:area-threshold=5000 \
-define connected-components:mean-color=true \
-connected-components 8 tmp.png | grep "gray(255)" | awk '{print $2}'`)
num=${#bboxArr[*]}
for ((i=0; i<num; i++)); do
bbox="${bboxArr[$i]}"
echo "$bbox;"
convert -quiet attrition_01.pdf +repage -alpha off -crop "$bbox" +repage boxes_$i.png
done

218x218+201+351;
214x210+351+122;
213x211+574+226;
203x214+583+6;
282x106+197+3;
197x146+591+450;
144x139+431+362;
158x159+23+423;
129x143+195+172;
140x158+25+17;
113x140+45+229;

```

[test1.zip](https://github.com/ImageMagick/ImageMagick/files/5197169/test1.zip)

**Reply by mwarning** [reply]
Awesome, I have to play a bit with the options to get a better feeling why this works, but the result is looking good.
Here is my shell conversion:

```
#!/bin/sh

input="$1"

if [ ! -f "$input" ]; then
  echo "File not found: $input"
  exit 1
fi

bboxArr=$(convert "$input" +repage -alpha off \
  -fuzz 15% -fill "rgb(235,193,134)" -opaque "rgb(235,193,134)" \
  +dither -colors 32   \
  -morphology close square:1 \
  -fuzz 0 -fill white +opaque "rgb(235,193,134)" -fill black +opaque white \
  -type bilevel \
  -define connected-components:exclude-header=true \
  -define connected-components:verbose=true \
  -define connected-components:area-threshold=5000 \
  -define connected-components:mean-color=true \
  -connected-components 8 mask.png | awk '/gray\(255\)/{print $2}')

i=0
for bbox in $bboxArr; do
  echo "$bbox"
  convert -quiet "$input" +repage -alpha off -crop "$bbox" +repage "bbox_$i.png"
  i=$((i + 1))
done
```

**Reply by fmw42** [reply]
You may want to have arguments for the background color in the image and for the area-threshold for filtering out small regions that are not your main sticker images.

You can put in +write tmpX.png for X=0,1,2... at different places to see what is being done there in the processing.

**Reply by mwarning** [reply]
ok, thanks. I currently try to figure out how to set the background.

**Reply by fmw42** [reply]
You could do the color reduction, then look at the histogram to get the most frequent color. Then continue the process from there.

**Reply by mwarning** [reply]
That is a bit too hard for me right now. It will take some time.

**Reply by fmw42** [reply]
The -fuzz 15% allows for variation in the background color. So you could just get the color from a corner (say, pixel 0,0) and use that, if the corner is reliably background. 

`color=$(convert image -format "%[pixel:u.p{0,0}]" info:)
`

**Reply by mwarning** [reply]
This seems to extract the background color:
```
background_color=$(convert "$input" -colors 32 -depth 8 -format "%c" histogram:info: | sort -r -n |  awk '{print($4); exit;}')
```

But it still fails for the example on stackoverflow. Probably because the background is white..

**Reply by fmw42** [reply]
Using the image on stackoverflow, your command works for me:

```
convert "T2VEJ.jpg" -colors 32 -depth 8 -format "%c" histogram:info: | sort -r -n |  awk '{print($4); exit;}'
srgb(252,252,251)
```

**Reply by mwarning** [reply]
For me it is:
```
$ convert "T2VEJ.jpg" -colors 32 -depth 8 -format "%c" histogram:info: | sort -r -n |  awk '{print($4); exit;}'
white
```
Which seems to make quite the difference. It works somewhat when I hard code "set srgb(252,252,251)", but I only get 5 objects, not ~27 as on stackoverflow. Objects also appear in other bboxes.

@fmw42 how do I get a different gray scale value for each object as on stackoverflow? I try to adapt their approach, as it does not just cut out objects by bbox, but by a mask. They also set the background transparent (which I like to have achieve as well).

**Comment by fmw42** [comment]
Use the code for connected components that he used. That is remove the line for mean-color=true (or set it to false) and add -auto-level at the end. See documentation at https://imagemagick.org/script/connected-components.php

```
convert T2VEJ.jpg -threshold 98% \
    -morphology dilate octagon                            \
    -define connected-components:area-threshold=800       \
    -define connected-components:verbose=true             \
    -connected-components 8 -auto-level PNG8:lumps.png
```

**Reply by mwarning** [reply]
My current approach. The edges are not as that nice as I would have used the fuzzy select in gimp, but it looks almost ok. 

it works for attrition_01.pdf and fails for T2VEJ.jpg. But that probably means that I need find out some automatic adjustment for the parameters.

```
#!/bin/sh

input="$1"

# verbose
set -x

if [ ! -f "$input" ]; then
  echo "File not found: $input"
  exit 1
fi

background_color=$(convert "$input" -colors 32 -depth 8 -format "%c" histogram:info: | sort -r -n |  awk '{print($4); exit;}')

# avoid white
if [ "$background_color" = "white" ]; then
  background_color="rgb(254,254,254)"
fi

echo "background_color: $background_color"

convert "$input" +repage -alpha off \
  -fuzz 5% -fill "$background_color" -opaque "$background_color" \
  +dither -colors 32   \
  -morphology dilate octagon \
  -fuzz 0 -fill white +opaque "$background_color" -fill black +opaque white \
  -type bilevel \
  -define connected-components:exclude-header=true \
  -define connected-components:verbose=true \
  -define connected-components:area-threshold=5000 \
  -connected-components 8 -auto-level lumps.png

greys=$(convert lumps.png -colors 255 -depth 8 -format "%c" histogram:info: | awk '!/#000000/{print($3)}')
mask=0
for grey in $greys; do
  convert lumps.png -fuzz 1% -fill black +opaque "$grey" -fill white -opaque "$grey" mask_${mask}.png
  mask=$((mask + 1))
done

seg=0
rm -f segment_*png
for f in mask_*png; do
   convert "$input" "$f" -compose copy-opacity -composite -trim +repage segment_${seg}.png
   seg=$((seg + 1))
done
```
EDIT: Some speed improvement & cleanup

**Reply by fmw42** [reply]
You can antialias the edges, if you use `-blur 0x2 -level 50x100%` on the mask. Change the 0x2 as desired

**Reply by mwarning** [reply]
That does not really address the problem. I will look further when I have time.

But a big thank you for your help so far! :-)
